shader-nodes/shader-nodes/flexOp.js

```diff
@@ -0,0 +1,65 @@
+// utility
+export var type2Comp = { float: 1, vec2: 2, vec3: 3, vec4: 4 };
+export var comp2Type = ['', 'float', 'vec2', 'vec3', 'vec4'];
+
+var flexOp = {
+    getRetType: function ( argTypes ) {
+        var maxComp = 0;
+        for (var argIndex = 0; argIndex < argTypes.length; argIndex++) {
+            maxComp = Math.max(maxComp, type2Comp[argTypes[argIndex]]);
+        }
+        return comp2Type[maxComp];
+    },
+    getHeadCode: function (retType, opName, argTypes, precision) {
+        precision = precision ? `_${precision}` : '';
+        var code = `${retType} ${opName}${precision}( `; // alter name to create variant
+        for (var argIndex = 0; argIndex < argTypes.length; argIndex++) {
+            code += `in ${argTypes[argIndex]} arg${argIndex}`;
+
+            code += (argIndex === argTypes.length - 1) ? ' )\n' : ', ';
+        }
+        return code;
+    },
+    genOp: function (opName, opCode, label ) {
+        var op = {
+            // placeholder for meta data
+            // placeholder for editor data
+            editor: { label: label },
+            // generator
+            gen: function ( argTypes, options ) {
+                return flexOp.gen(argTypes, options, opName, opCode );
+            }
+        };
+        return op;
+    },
+    gen: function ( argTypes, options, opName, opCode) {
+        // determine return type
+        var retType = this.getRetType(argTypes);
+
+        // construct head code
+        var code = this.getHeadCode(retType, opName, argTypes, options ? options.precision : '');
+
+        // construct body code
+        code += '{\n';
+
+        code += '    return ';
+
+        for (var argIndex = 0; argIndex < argTypes.length; argIndex++) {
+            // convert all arguments to return type
+            code += `${retType}(arg${argIndex}`;
+
+            // expand if needed - extra components are set to zero - NB scalar values are broadcast to all components
+            var argComp = type2Comp[argTypes[argIndex]];
+            var extend = type2Comp[retType] - argComp;
+            code += (extend > 0 && argComp !== 1) ? `, ${comp2Type[expand]}(0))` : ')';
+
+            code += (argIndex === argTypes.length - 1) ? ';\n' : ` ${opCode} `;
```

Comment:
it feels like this should be on a higher level when the generated node function gets called .. so that it works any node parameters (and not only flex nodes), without having to implement this for each core node / subgraph?

Reply:
Auto type conversion actually already does work on all nodes in the graph - in combination with (an optional) per connection swizzle - which makes it very flexible - so you can feed a scalar value into vector port or vice versa or a vec2 into a vec3 or a vec4 into a vec3 etc and it gets automatically extended or truncated: https://github.com/playcanvas/engine/blob/3b7e6dbf40c1f1e1a844234e5dd9579da712cf7e/src/scene/materials/shader-graph-node.js#L423 - its pretty easy to try in the engine examples - I can also add one that demos Auto type conversion? It should also become more intuitive when implemented in the graph editor with colors (like in Lens Studio) and/or numbers (link in Unity) indicating the type conversion.

The flexOp is actually distinct from general auto type conversion because it and can take a flexible number of arguments of any (scalar or vector) type and changes the type of the operation itself and based on the argument with the largest number of components - this means other arguments may need to be extended.